pkg/persistence/rethinkdb/store.go

```diff
@@ -0,0 +1,99 @@
+package rethinkdb
+
+import (
+	"strings"
+	"time"
+
+	"github.com/aquasecurity/harbor-scanner-trivy/pkg/etc"
+	"github.com/aquasecurity/harbor-scanner-trivy/pkg/harbor"
+	"github.com/aquasecurity/harbor-scanner-trivy/pkg/job"
+	"github.com/aquasecurity/harbor-scanner-trivy/pkg/persistence"
+
+	"github.com/sirupsen/logrus"
+	"golang.org/x/xerrors"
+	r "gopkg.in/rethinkdb/rethinkdb-go.v6"
+)
+
+type rethinkdbScanJob struct {
+	LastUpdatedAt time.Time `json:"last_updated_at"`
+	job.ScanJob
+}
+
+type store struct {
+	conn r.QueryExecutor
+	cfg  etc.Rethink
+}
+
+func NewStore(conn r.QueryExecutor, cfg etc.Rethink) persistence.Store {
+	return &store{
+		conn: conn,
+		cfg:  cfg,
+	}
+}
+
+func (s *store) Create(scanJob job.ScanJob) error {
+	logrus.WithFields(logrus.Fields{
+		"scan_job_id":     scanJob.ID,
+		"scan_job_status": scanJob.Status.String(),
+		"rethinkdb_table": s.cfg.ScansTable,
+	}).Debug("Saving scan job")
+
+	rsj := rethinkdbScanJob{
+		LastUpdatedAt: time.Now().UTC(),
+		ScanJob:    scanJob,
+	}
+
+	res, err := r.Table(s.cfg.ScansTable).Insert(rsj).RunWrite(s.conn)
+	if err != nil || res.Inserted != 1 {
+		return xerrors.Errorf("inserting scan job failed: %w", err)
+	}
+
+	return nil
+}
+
+func (s *store) Get(scanJobID string) (*job.ScanJob, error) {
+	scanJob := new(job.ScanJob)
+	if err := r.Table(s.cfg.ScansTable).Get(scanJobID).ReadOne(scanJob, s.conn); err != nil {
+		if err == r.ErrEmptyResult {
+			return nil, nil
+		}
+
+		return nil, xerrors.Errorf("getting scan job failed: %w", err)
+	}
+
+	return scanJob, nil
+}
+
+func (s *store) UpdateStatus(scanJobID string, newStatus job.ScanJobStatus, errors ...string) error {
+	logrus.WithFields(logrus.Fields{
+		"scan_job_id": scanJobID,
+		"new_status":  newStatus.String(),
+	}).Debug("Updating status for scan job")
+
+	res, err := r.Table(s.cfg.ScansTable).Get(scanJobID).Update(map[interface{}]interface{}{
```

Comment:
Similar to https://github.com/hrist0stoichev/harbor-scanner-trivy/blob/RethinkDb/pkg/persistence/redis/store.go#L109 would probably use a data structure for the scan job

Reply:
Same question as above.

Reply:
Same as above, but at this point we don't have the status, and without marking it `omitempty`, it likely will set that string to an empty string.

Leaves the options of marking it as `omitempty` for the scanjob status, or to query and then write, or to create a private struct which also seems reasonable 

`type reportUpdate struct {
     ID     string            `json:"id"`
     Report *harbor.ScanReport `json:"report,omitempty"`
     LastUpdatedAt time.Time `json:"last_updated_at"`
}

func NewReportUpdate(id string, report *harbor.ScanReport) *reportUpdate {
    return &reportUpdate {
         ID: id,
         Report: report,
         LastUpdatedAt: time.Now(),
    }
}
`

Reply:
Ideally we should reuse core domain struct and implement external adapters such as the one for Redis or Rething DB. I know that it might require some changes in the core interfaces and we don't have to do that in this PR, but what I'd like to have at some point is this model hex architecture:

![hex-harbor-scanner-trivy-redis](https://user-images.githubusercontent.com/1322923/82072618-48ff1380-96d8-11ea-9a50-435386bceff1.png)

![hex-harbor-scanner-trivy-rethinkdb](https://user-images.githubusercontent.com/1322923/82072639-4e5c5e00-96d8-11ea-88ba-a7ddc7ffc599.png)

Anyway, it's great contribution and I'll also review the whole code ASAP @cquon @hrist0stoichev

Reply:
@danielpacak I don't really understand the diagrams that you've attached. Can you elaborate a bit further? I'm already using the internal Trivy-Adapter structures but I need to add some fields that are missing and this is the reason why I create private structs that extend the Trivy-Adapter ones. By field that are missing I mean fields that are only necessary for the RethinkDB implementation to work and are not needed elsewhere. For example, Redis has TTL built in but RethinkDB does not. That's why I need to add a timestamp field so that I know when the TTL has expired and I can delete the database record. For other database implementation, other fields might be needed and so on.